backend/session.go

```diff
@@ -0,0 +1,173 @@
+/*
+AUTHORS
+  David Sutton <davidsutton@ausocean.org>
+
+LICENSE
+  Copyright (C) 2024 the Australian Ocean Lab (AusOcean).
+
+  This is free software: you can redistribute it and/or modify it
+  under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This is distributed in the hope that it will be useful, but WITHOUT
+  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+  License for more details.
+
+  You should have received a copy of the GNU General Public License
+  in gpl.txt. If not, see http://www.gnu.org/licenses/.
+*/
+
+package backend
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"reflect"
+	"time"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/gorilla/sessions"
+)
+
+// Session defines an interface for a session to keep track of user
+// authenticated sessions.
+type Session interface {
+	// SetMaxAge sets the Max Age of the session, after which the session is
+	// no longer valid.
+	SetMaxAge(age time.Duration) error
+
+	// Set sets a key value store in the session.
+	Set(key string, value any) error
+
+	// Get retrieves the value for a given key in the session and stores it in the destination.
+	Get(key string, dst any) error
+
+	// Invalidate immediately invalidates the session and marks it as no
+	// longer valid.
+	Invalidate() error
+}
+
+// FiberSession implements the Session interface using a Fiber Cookie based
+// storage method.
+type FiberSession struct {
+	cookie *fiber.Cookie              // Cookie used to store the session.
+	values map[string]json.RawMessage // Map of the key value pairs to be encoded into the session.
+}
+
+// NewFiberSession creates a new empty FiberSession with the given id.
+func NewFiberSession(id, value string) (*FiberSession, error) {
+	s := &FiberSession{cookie: &fiber.Cookie{Name: id}, values: make(map[string]json.RawMessage)}
+
+	if value == "" {
+		return s, nil
+	}
+
+	// Parse the value into the session value map.
+	ckValue, err := url.QueryUnescape(value)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unescape cookie value: %w", err)
+	}
+	err = json.Unmarshal([]byte(ckValue), &s.values)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal value: %w", err)
+	}
+
+	return s, nil
+}
+
+// SetMaxAge implements the SetMaxAge method of the Session interface by setting
+// the maximum age of the cookie.
+func (s *FiberSession) SetMaxAge(age time.Duration) error {
+	s.cookie.MaxAge = int(age.Seconds())
+	return nil
+}
+
+// Set implements the Set method of the Session interface by encoding a query escaped
+// map in JSON format to the cookie value.
+func (s *FiberSession) Set(key string, value any) error {
+	// Convert the value into a json RawMessage.
```

Comment:
Probably don't need this.

Reply:
Which bit sorry? just the comment or some of the logic?

Reply:
Just the comment